### PR TITLE
DataChannel termination via the new DirectTransport

### PR DESCRIPTION
**Pull request by ibc**
Related to #400

### Feature overview

- Allow the Node application to directly send text/binary messages to `mediasoup-worker` C++ process so others can consume them using `DataConsumer`.

- And vice-versa: allow the Node application to directly consume in Node messages send by `DataProducers`.

### Implementation

- New `DirectTransport` (created via `router.createDirectTransport`) that uses the new `PayloadChannel` (UnixSocket pair between Node and C++) to send payload messages in both directions.

- Extend `DataProducer` and `DataConsumer` classes to not just work with SCTP capable transports (`WebRtcTransport`, `PlainTransport`, `PipeTransport`) but also with the new `DirectTransport`.
  - In the latter, there are no "SCTP stream parameters" involved at all.

- `sctpStreamParameters` are now optional in `DataProducerOptions` (just valid for `directTransport.produceData()`).
  - So `dataProducer/Consumer.sctpStreamParameters` getter can now return `undefined`.

- Added `dataProducer/Consumer.type` getter, that can be 'sctp' or 'direct'.

- `transport.consumeData()` (when in an SCTP capable transport) now accepts optional SCTP related parameters in `DataConsumerOptions` to tell mediasoup how to deliver messages to the remote SCTP endpoint:

```ts
ordered?: boolean;
maxPacketLifeTime?: number;
maxRetransmits?: number;
```

- `Worker`: Added a new `message` log tag.

### CODE REVIEW

**Comment by ibc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- worker/include/RTC/WebRtcTransport.hpp  3
- test/test-DirectTransport.js  15
- test/test-WebRtcTransport.js  3
- worker/include/RTC/DataConsumer.hpp  1
- worker/include/RTC/PlainTransport.hpp  2
- lib/Channel.js  4
- lib/DirectTransport.js  1
- worker/src/RTC/DataConsumer.cpp  1
- worker/include/RTC/DirectTransport.hpp  3
- worker/include/RTC/PipeTransport.hpp  2
- lib/Router.js  1
- test/test-DataConsumer.js  1
- worker/include/RTC/DataProducer.hpp  1
- test/test-PlainTransport.js  6
- worker/src/Channel/UnixStreamSocket.cpp  4
- worker/src/RTC/DataProducer.cpp  1
- worker/include/Channel/UnixStreamSocket.hpp  2
- worker/include/PayloadChannel/UnixStreamSocket.hpp  2
- lib/PayloadChannel.js  5
- worker/src/PayloadChannel/UnixStreamSocket.cpp  4
         

Clones removed
==============
+ test/test-PipeTransport.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/versatica/mediasoup/pullRequest?prid=5604213&bid=18357577)

**Comment by ibc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- lib/Channel.js  4
- worker/include/RTC/PlainTransport.hpp  2
- worker/src/Channel/UnixStreamSocket.cpp  4
- worker/include/PayloadChannel/UnixStreamSocket.hpp  2
- worker/include/RTC/DirectTransport.hpp  3
- test/test-WebRtcTransport.js  3
- worker/include/RTC/DataProducer.hpp  1
- test/test-DataConsumer.js  1
- worker/src/RTC/DataProducer.cpp  1
- worker/src/PayloadChannel/UnixStreamSocket.cpp  4
- test/test-PlainTransport.js  6
- worker/include/RTC/DataConsumer.hpp  1
- lib/DirectTransport.js  1
- lib/PayloadChannel.js  5
- test/test-DirectTransport.js  15
- worker/include/Channel/UnixStreamSocket.hpp  2
- worker/include/RTC/PipeTransport.hpp  2
- worker/src/RTC/DataConsumer.cpp  1
- lib/Router.js  1
- worker/include/RTC/WebRtcTransport.hpp  3
         

Clones removed
==============
+ test/test-PipeTransport.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/versatica/mediasoup/pullRequest?prid=5604213&bid=18357577)

File: src/Worker.ts [hunk 1]
```diff
@@ -260,7 +264,9 @@ export class Worker extends EnhancedEventEmitter
 				// fd 2 (stderr)  : Same as stdout.
 				// fd 3 (channel) : Producer Channel fd.
 				// fd 4 (channel) : Consumer Channel fd.
-				stdio : [ 'ignore', 'pipe', 'pipe', 'pipe', 'pipe' ]
+				// fd 3 (channel) : Producer PayloadChannel fd.
```
Comment (jmillan) [inline src/Worker.ts]:
```suggestion
				// fd 5 (channel) : Producer PayloadChannel fd.
```

File: src/Worker.ts [hunk 2]
```diff
@@ -260,7 +264,9 @@ export class Worker extends EnhancedEventEmitter
...
+				// fd 4 (channel) : Consumer PayloadChannel fd.
```
Comment (jmillan) [inline src/Worker.ts]:
```suggestion
				// fd 6 (channel) : Consumer PayloadChannel fd.
```

File: src/DirectTransport.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	}
+
+	/**
+	 * @override
+	 */
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	async consumer(options: ConsumerOptions): Promise<Consumer>
+	{
+		throw new UnsupportedError('consumer() not implemented in DirectTransport');
+	}
```
Comment (jmillan) [inline src/DirectTransport.ts]:
```suggestion
	async consume(options: ConsumerOptions): Promise<Consumer>
	{
		throw new UnsupportedError('consume() not implemented in DirectTransport');
	}
```

File: worker/include/RTC/Transport.hpp [hunk 4]
```diff
@@ -116,9 +117,10 @@ namespace RTC
 		// Subclasses must also invoke the parent Close().
 		virtual void FillJson(json& jsonObject) const;
 		virtual void FillJsonStats(json& jsonArray);
-		// Subclasses must implement this method and call the parent's one to
+		// Subclasses must implement these method and call the parent's one to
```
Comment (jmillan) [inline worker/include/RTC/Transport.hpp]:
```suggestion
		// Subclasses must implement these methods and call the parent's ones to
```

File: worker/src/PayloadChannel/Notification.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+namespace PayloadChannel
+{
+	/* Class variables. */
+
+	// clang-format off
+		std::unordered_map<std::string, Notification::EventId> Notification::string2EventId =
+		{
+			{ "dataProducer.send", Notification::EventId::DATA_PRODUCER_SEND }
+		};
+	// clang-format on
```
Comment (jmillan) [inline worker/src/PayloadChannel/Notification.cpp]:
```suggestion
	// clang-format off
	std::unordered_map<std::string, Notification::EventId> Notification::string2EventId =
	{
		{ "dataProducer.send", Notification::EventId::DATA_PRODUCER_SEND }
	};
	// clang-format on
```

File: worker/src/RTC/Transport.cpp [hunk 6]
```diff
@@ -34,6 +35,34 @@ namespace RTC
 	{
 		MS_TRACE();
 
+		auto jsonDirectIt = data.find("direct");
+
+		// clang-format off
+		if (
+			jsonDirectIt != data.end() &&
+			jsonDirectIt->is_boolean() &&
+			jsonDirectIt->get<bool>()
```
Comment (jmillan) [inline worker/src/RTC/Transport.cpp:44]: isn't is redundant?

**Review by jmillan — APPROVED** [review]
masterpiece :+1:

File: worker/src/RTC/Transport.cpp (hunk 6 above)
Comment (ibc) [reply worker/src/RTC/Transport.cpp:44]: Because we want to check if it's value is true, not just if it's  a boolean.

Comment (jmillan) [reply worker/src/RTC/Transport.cpp:44]: right  :+1:

File: worker/src/RTC/Transport.cpp [hunk 7]
```diff
@@ -947,13 +985,13 @@ namespace RTC
...
 					MS_THROW_ERROR("SCTP not enabled");
 
 				std::string dataProducerId;
 
 				// This may throw.
-				SetNewDataProducerIdFromRequest(request, dataProducerId);
+				SetNewDataProducerIdFromInternal(request->internal, dataProducerId);
 
 				// This may throw.
 				auto* dataProducer = new RTC::DataProducer(dataProducerId, this, request->data);
```
Comment (jmillan) [inline worker/src/RTC/Transport.cpp:1002]: Verify that producer type corresponds with transport type.

File: worker/src/RTC/Transport.cpp [hunk 8]
```diff
@@ -1002,27 +1040,23 @@ namespace RTC
...
+				SetNewDataConsumerIdFromInternal(request->internal, dataConsumerId);
 
 				// This may throw.
 				auto* dataConsumer = new RTC::DataConsumer(
-				  dataConsumerId,
-				  dataProducerId,
-				  this,
-				  request->data,
-				  this->sctpAssociation->GetMaxSctpMessageSize());
+				  dataConsumerId, dataProducerId, this, request->data, this->maxMessageSize);
```
Comment (jmillan) [inline worker/src/RTC/Transport.cpp:1107]: Verify that consumer type corresponds with transport type.

File: worker/src/RTC/Transport.cpp (hunk 7 above)
Comment (ibc) [reply worker/src/RTC/Transport.cpp:1002]: Done.

**Comment by ibc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- worker/src/RTC/DataConsumer.cpp  1
- test/test-PlainTransport.js  6
- worker/src/RTC/DataProducer.cpp  1
- worker/include/Channel/UnixStreamSocket.hpp  2
- lib/PayloadChannel.js  5
- worker/include/RTC/DataConsumer.hpp  1
- worker/src/Channel/UnixStreamSocket.cpp  4
- lib/Router.js  1
- worker/include/RTC/WebRtcTransport.hpp  3
- lib/DirectTransport.js  1
- worker/src/PayloadChannel/UnixStreamSocket.cpp  4
- test/test-node-sctp.js  1
- worker/include/RTC/DirectTransport.hpp  3
- test/test-DataConsumer.js  1
- worker/include/RTC/PlainTransport.hpp  2
- test/test-DirectTransport.js  15
- worker/include/RTC/DataProducer.hpp  1
- worker/include/RTC/PipeTransport.hpp  2
- worker/include/PayloadChannel/UnixStreamSocket.hpp  2
- lib/Channel.js  4
- test/test-WebRtcTransport.js  3
         

Clones removed
==============
+ test/test-PipeTransport.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/versatica/mediasoup/pullRequest?prid=5604213&bid=18357577)

**Comment by ibc** [comment]
Here the announcement:

https://mediasoup.discourse.group/t/mediasoup-3-6-0-is-out-with-datachannel-termination-in-node-js/1118

**Pull request merged**